[DPE-7377] Refresh V3

Batalex opened the pull request:
This PR implements refresh v3. Please do have a look at the notes below. There are some design decisions there you might have an opinion on.

## Changes and design decisions

- Add refresh v3 (I'll call it V3 from now on). This change has some pretty big impacts in multiple places:
     - We can scrape everything we had in `events/upgrades.py` 
     - We can remove our logic about having a trusted app.  V3 is executed before any event handler and already checks if the app is trusted, so our own logic was redundant.
- Enforce that the workload image and the job image use the same spark version through a unit test. We test this down to the patch version. 
     - For now, I made a choice to ignore the patch version altogether. Do we agree on that, or do we need to implement additional logic there as well? (e.g. prevent patch downgrade)
     - This test makes some external call to fetch the images' metadata. I don't think it is necessary to have a pytest mark to skip those (is it useful to run the full unit test suite without connectivity?)
     - We check for a full spark version match between the workload and the job image. If we want to, we can add some more logic to enable a 3.4.4 workload running 3.4.2 jobs.
- Reword "in-place upgrades" to "refresh". Since we have a dedicated term for it, we might as well use it.
- Bonus: Solved the annoying error log `This operation (fetch_my_relation_data()) can only be performed by the leader unit`

### Integration tests
V3 with multiple units can only refresh to another revision implementing V3.
Therefore, I released revision 87 from 42cc29c2c9b4d46d139f4fb4cd61ba40b39a109f to the `latest/edge/refresh` branch.
 You can see refresh integration tests against this revision here: https://github.com/canonical/kyuubi-k8s-operator/pull/87.

### Steps before/after merging

1. We need to make sure that we keep working on `latest` before GA, so change the 3.4 value in `refresh_versions.toml`
1. Disable refresh integration tests here; uncomment the line in `charmcraft.yaml` and remove `charm` key in `refresh_versions.toml` so that we get a new revision on edge
1. Re-enable refresh integration test
1. Once we go GA:
    1. Create the 3.4 track on charmhub (+ 3.5). Rename the main branch to 3.4/edge?
    2. Change the latest track in the `tag` job and in `refresh_versions.toml` to 3.4
    3. If multiple tracks, create a new branch and repeat the step above for 3.5. Adapt the release CI job to target this new branch as well.
1. Anything else?

Batalex committed:
ci: Upgrade shared workflows to v32

Batalex committed:
feat: Initial support inplace upgrade v3

Batalex committed:
feat: Add charm-refresh-build-version logic to charmcraft.yaml

Batalex committed:
tests: Fix missing field in refresh_versions.toml in unit tests

Batalex committed:
feat: Initial is_compatible() implementation

Batalex committed:
WIP: Disable write-charm-version

Batalex committed:
fix: Fix lightkube permission error

Batalex committed:
refactor: Remove redundant trust logic

Batalex committed:
refactor: Rework inplace upgrade as refresh

Batalex committed:
chore: Update job image to spark 3.4.4

Batalex committed:
tests: Add spark image compatibility unit test

Batalex committed:
tests: Switch integration trust tests for ease

Batalex committed:
refactor: Extract workload compatibility logic into its own function

Batalex committed:
tests: Adapt refresh integration tests to V3 paradigm

Batalex committed:
tests: Add integration test for failed refresh & rollback

Batalex committed:
feat: Add pre refresh checks

Batalex committed:
fix: Avoid non leader access errors

Batalex committed:
feat: Wire refresh statuses in compute_status decorator

Batalex committed:
fix: Use local folder for faulty charm

Batalex committed:
tests: Reduce update status interval for faster int. tests

Batalex committed:
fix: Reconcile status after unit refreshes

Batalex committed:
WIP: Wrap resume-refresh task failure?

Batalex committed:
fix: Fix faulty charm path in refresh call

Batalex committed:
fix: Fix test assertion in rollback int. test

Batalex committed:
fix: Wait for agent idle in rollback int. test

Batalex committed:
fix: Fix refresh int. test with image upgrade

Batalex commented on .github/workflows/release.yaml:
```diff
@@ -1,28 +1,39 @@
 on:
   push:
     branches:
       - main
```
**todo**: add `3.4/edge` and `3.5/edge`

Batalex commented on .github/workflows/release.yaml:
```diff
@@ -1,28 +1,39 @@
+    name: Create charm refresh compatibility version git tag
+    uses: canonical/data-platform-workflows/.github/workflows/tag_charm_edge.yaml@v32.0.0
+    with:
+      # TODO(ga): Update to track 1/
+      track: latest
```
todo: use 3.4 on when targeting 3.4/edge (resp. 3.5)

Batalex committed:
chore: Remove former upgrade lib

Batalex committed:
chore: Reword job oci image constant

Batalex committed:
ci: Prepare CI for multi branches dev flow

Batalex commented on charmcraft.yaml:
```diff
@@ -82,8 +82,23 @@ parts:
+      source ../refresh-version-venv/bin/activate
+      poetry install --only build-refresh-version
+      # FIXME: uncomment before merging
+      # write-charm-version
```
todo

Batalex commented on src/events/refresh.py:
```diff
@@ -0,0 +1,121 @@
+            )
+
+        if planned_units == 1:
+            logger.warning(
+                "Refreshing a single unit deployment is not recommended, "
+                "as compatibility check will not be run."
+            )
```
**question**: Is there anyway to log this in the action output?

Batalex commented on tests/integration/refresh/test_refresh.py:
```diff
@@ -0,0 +1,279 @@
+    try:
+        juju.run(leader_unit, "resume-refresh")
+    except jubilant.TaskError:
+        # TODO: Ask about this
+        # Not sure why we get a task terminated while things are apparently going fine.
```
This is a juju bug, see https://github.com/canonical/charm-refresh/blob/4c82e341084b50443144511b1ca40218bcaa6165/charm_refresh/_main.py#L1872-L1874.

**todo**: document here

deusebio reviewed (commented):
Very good job. I have raise a few points, nothing really major but worth clarifying before approving. Also, I'm not sure I can really be very competent to review how upgrade v3 is integrated here, as I probably miss context, ~~but I'm thinking that perhaps it would be worth asking Carl a quick review on this~~ (sorry just realized Carl is among the reviewers already)

Batalex commented on src/events/refresh.py:
```diff
@@ -0,0 +1,121 @@
+        logger.error(
+            "Upgrading to a previous minor workload is not supported. "
+            f"Got {old_major}.{old_minor} to {new_major}.{new_minor}"
+        )
```
This check is here to prevent downgrades, not rollbacks. Semantics aside, this is my interpretation of the spec for refresh v3. Since most products do not support forward compatibility, downgrading is usually a no-go.
However, since Kyuubi is mostly stateless and seems to stay on the same Hive version, maybe we can loosen this constraint.
What do you think?

Batalex commented on src/events/base.py:
```diff
@@ -30,23 +30,28 @@ class BaseEventHandler(Object, WithLogging):
-    def get_app_status(  # noqa: C901
-        self,
+    def get_app_status(  # noqa: C901 - ignore complexity threshold
+        self, check_refresh: bool = False
```
This makes more sense when we look at it from the perspective of the `@compute_status` decorator; but there's probably a better way to implement that.
Since we use the same function to compute the app and the unit's statuses, we needed a way to not set the "unit refresh status" as the application one.
Maybe it's time for us to introduce two separate functions for that

Batalex replied:
Added to the docstrings in af9c4a94f8522cf13439574995cd43fac05df9cb.

Batalex committed:
feat: Improve status handling while refreshing

Batalex committed:
docs: Document juju bug in refresh int. tests

welpaolo reviewed (approved):
Great work Alex! I left some minor comments.

Batalex commented on src/events/base.py:
```diff
@@ -30,23 +30,28 @@ class BaseEventHandler(Object, WithLogging):
+            workload=self.workload,
         )
 
-        if not k8s_manager.has_cluster_permissions():
```
`charm-refresh` checks for cluster permission directly in the `__init__` before any other event, so our logic is redundant

Batalex committed:
chore: Bump src/events/refresh.py date

Co-authored-by: welpaolo <paolo.sottovia@canonical.com>

Batalex commented on src/events/refresh.py:
```diff
@@ -0,0 +1,121 @@
+
+    _charm: KyuubiCharm
+
+    def run_pre_refresh_checks_after_1_unit_refreshed(self) -> None:
```
No, a missing auth db does not prevent the refresh right now but we can make having one mandatory. One immediate benefit would be that since we don't start the workload without one, we could wire that logic in refresh v3 and only refresh the next unit on the list if the service is running

Batalex committed:
ci: Disable refresh integration tests

Batalex committed:
tests: Fix fetch_connection_info in rollback test after rebase

Batalex committed:
docs: Document statuses priority with refresh v3

Batalex committed:
chore: Enable tag based build

Batalex commented on src/events/base.py:
```diff
@@ -83,6 +88,13 @@ def get_app_status(  # noqa: C901
+        if (
+            check_refresh
+            and self.charm.refresh is not None
+            and (refresh_status := self.charm.refresh.unit_status_lower_priority()) is not None
```
Documented this a bit more in 49a37026ce516cf2814fd89fe75149c71630d691

carlcsaposs-canonical reviewed (commented):
refresh v3 requires that the charm check .workload_allowed_to_start before starting the workload. more info: https://github.com/canonical/charm-refresh/blob/a5b93631c6019faa3bb0153b14385f9ca41111a2/charm_refresh/_main.py#L543-L583

this is to prevent data loss & ensure rollback is possible if the compatibility checks, pre-refresh-checks, or workload container compatibility checks fail

---

are there any actions that should not happen or could be dangerous while a refresh is in progress? for some examples, see both sheets of https://docs.google.com/spreadsheets/d/1_qjINUoj5Mi7PlXhCdpwQ0170bIZSgpvNFMAnTOtuQw/edit?gid=1617626956#gid=1617626956. I noticed that .in_progress isn't being checked anywhere

---

fyi, didn't review tests closely

Batalex commented on src/events/actions.py:
```diff
@@ -97,10 +97,6 @@ def _on_set_password(self, event: ActionEvent) -> None:
                 lambda: self.get_app_status() != Status.ACTIVE.value,
                 "The action failed because the charm is not in active state.",
             ),
-            (
-                lambda: not self.charm.upgrade_events.idle,
-                f"Cannot set password while upgrading (upgrade_stack: {self.charm.upgrade_events.upgrade_stack})",
-            ),
```
Yes, however, since those actions are about to be removed in the coming days, I did not pay them too much attention.

deusebio replied:
Yes, totally! let's not spend time on things we are going to remove, reducing the efficiency now, and later to remove more code :joy:

Batalex commented on src/charm.py:
```diff
@@ -90,6 +86,29 @@ def __init__(self, *args):
+        except (charm_refresh.UnitTearingDown, charm_refresh.PeerRelationNotReady):
+            self.refresh = None
+
+        if (
+            self.refresh is not None
+            and not self.refresh.next_unit_allowed_to_refresh
+            and self.refresh.workload_allowed_to_start
+        ):
+            self.refresh.next_unit_allowed_to_refresh = True
```
We don't have anything yet to check that the unit is healthy. The closest thing we have is that if we don't have an authentication db related the workload does not start. But having one (which starts the pebble service) does not mean that it's healthy
@deusebio @welpaolo not sure what we can do there

Batalex replied:
Let's circle back to https://github.com/canonical/kyuubi-k8s-operator/pull/85#discussion_r2135764261. If we make it a requirement to have the auth-db related before we can refresh, then no problem, we can check workload service

deusebio replied:
> even if you don't make auth-db a requirement, I would recommend checking the workload service if auth-db is present

Seems sensible to me. Also I'm thinking that if we don't have other required information, we should probably not start the service as well. This is a different ticket, but having the check for auth-db would create the structure to extent to other cases where if the pre-conditions are not met, we don't check, otherwise we make sure the workload is "healthy" (I understand that Pebble is just a simple check, but better than nothing)

Batalex replied:
Added check in bf38eac53ba3634fff28a4cd59bd3300d3c49343. We are also looking for a better health test with Paolo, but that would be out of scope of this PR

Batalex commented on src/charm.py:
```diff
@@ -90,6 +86,29 @@ def __init__(self, *args):
+                    _charm=self,
+                )
+            )
+        except (charm_refresh.UnitTearingDown, charm_refresh.PeerRelationNotReady):
+            self.refresh = None
```
Hum, I might be missing some logic somewhere because the charm cannot start without this except. But it's also easier to just prevent refreshing while the cluster is unstable, and we have such a check for that in pre-refresh-check. I'll keep it as-is for the time being

Batalex replied:
Well then, seems like I did not get the point as well as I thought after all.
But I stand by my original comment, I cannot think of a single dangerous operation. Even the metastore can be re-created if needed with minimal impact on the operations.
Let's keep this thread open in case @deusebio, @welpaolo or @theoctober19th would like to chime in

deusebio replied:
If this counts as a dangerous operation (but being stateless I don't think it would be THAT dangerous), I'm fine to NOT provide support for scaling up and down during refresh. I'd rather say that the human operator should do this either before or after refresh, not during. We can also put a disclaimer in the docs.

Batalex replied:
This PR has already massively grown, let's go with the disclaimer approach for now

Batalex committed:
refactor: Remove decorator based status setting

Batalex committed:
docs: Apply suggestions

Batalex committed:
chore: Add fixme comment about track comparison

Batalex commented on src/events/refresh.py:
```diff
@@ -0,0 +1,121 @@
+                "Application is not related to an external metastore. "
+                "Refreshing will lead to local data loss."
+            )
+        elif not self._charm.metastore_events.metastore_manager.is_metastore_valid():
+            raise charm_refresh.PrecheckFailed("Metastore is invalid")
```
This command can be executed as long as the workload container is reachable

Batalex committed:
chore: Improve logging configuration

Batalex commented on src/charm.py:
```diff
@@ -44,7 +46,7 @@ class KyuubiCharm(TypedCharmBase[CharmConfig]):
 
-    def __init__(self, *args):
+    def __init__(self, *args) -> None:
         super().__init__(*args)
```
Done in 34c07261de413baf0a8577fce0cfee8fc894f293. I realized I did it outside the class, but that should not be a problem

Batalex commented on src/events/refresh.py:
```diff
@@ -0,0 +1,121 @@
+                "Refreshing will lead to local data loss."
+            )
+        elif not self._charm.metastore_events.metastore_manager.is_metastore_valid():
+            raise charm_refresh.PrecheckFailed("Metastore is invalid")
```
Reworded as part of f6e1084e4ad1ce2230987ea1e4a7cc389d9f126d

Batalex commented on tests/integration/refresh/test_refresh.py:
```diff
@@ -0,0 +1,287 @@
+        juju.run(leader_unit, "resume-refresh")
+    except jubilant.TaskError:
+        # Juju bug
+        # see https://github.com/canonical/charm-refresh/blob/4c82e341084b50443144511b1ca40218bcaa6165/charm_refresh/_main.py#L1872-L1874
```
Reworded in f6e1084e4ad1ce2230987ea1e4a7cc389d9f126d

Batalex commented on src/charm.py:
```diff
@@ -90,6 +86,29 @@ def __init__(self, *args):
+            self.refresh.next_unit_allowed_to_refresh = True
+
+        # Reconcile app and unit status
+        self.on.update_status.emit()
```
Done in 1849be5969898a8f040013655665e2c94240c2e5

Batalex committed:
chore: Lint & move logger config

Batalex commented on src/events/refresh.py:
```diff
@@ -0,0 +1,124 @@
+        """
+        # FIXME(refresh): Remove once this check is upstreamed
+        if not old_charm_version.track == new_charm_version.track:
+            logger.error(
+                "Refreshing to a different track is not supported. "
+                f"Got {old_charm_version.track} to {new_charm_version.track}"
+            )
+            return False
```
Can we find a middle ground and set to warning? Even if it's recoverable, it's still something that requires the user's manual intervention to fix

Batalex replied:
Alright, downgraded levels in ee2f2f027bfb1f5fb46f75bf0aa0f16ef0cf0d2d

Batalex commented on tests/integration/refresh/test_refresh.py:
```diff
@@ -0,0 +1,287 @@
+    task_params = {"check-compatibility": False}
+
+    if not with_image_upgrade:
+        assert (
+            "Refresh incompatible"
+            in status.apps[APP_NAME].units[refresh_order[0]].workload_status.message
+        ), "Application refresh not blocked due to incompatibility"
```
Oh, yes you are right. I did not understand what you meant. I'll rework that

Batalex replied:
Initial rework here: deeaf776df1f5c88e4e427f398304c44c1e6e12a
I'll publish new temp revisions to see if everything works as expected before closing the thread

Batalex replied:
Tests are more-or-less working at #87. Still need to fix a few minor things, but the refresh itself is working as advertised

Batalex committed:
refactor: Downgrade refresh logs to info level

Batalex committed:
tests: Adapt refresh int tests for running on main branches

Batalex committed:
docs: Document status setting in collect-* event handlers

Batalex committed:
feat: Check active workload before allowing next unit refresh

Batalex committed:
chore: Remove temp false assertion

Batalex committed:
tests: Fix rollback integration test

Batalex commented:
Addressed all comments, tests are passing on #87 :tada:

carlcsaposs-canonical reviewed (commented):
> Addressed all comments

have the two comments here been addressed? https://github.com/canonical/kyuubi-k8s-operator/pull/85#pullrequestreview-2921484254 (workload_allowed_to_start, in_progress)

Batalex committed:
tests: Only disable workload check on image upgrade

Batalex committed:
feat: Check workload allowed to start in kyuubi manager

Batalex commented on tests/integration/refresh/test_refresh.py:
```diff
@@ -0,0 +1,292 @@
+        task_params = {"check-compatibility": False}
+
+    if "missing/incorrect OCI resource" in first_unit_status_msg:
+        task_params = {"check-compatibility": False, "check-workload-container": False}
```
Reworked in 8fb1693f4dd99b0e6f1315b48021ef07f9669e07

Batalex commented:
> > Addressed all comments
> 
> have the two comments here been addressed? [#85 (review)](https://github.com/canonical/kyuubi-k8s-operator/pull/85#pullrequestreview-2921484254) (workload_allowed_to_start, in_progress)

For the second point, we settled on documenting what is supported / not.
Concerning the first:
- bf38eac53ba3634fff28a4cd59bd3300d3c49343 Checks that the workload is active before allowing the next unit to refresh
- eb0b99d594b6295593bd4ac98748a1d059148d4a Checks `workload_allowed_to_start` on every event handler that would start the workload

@carlcsaposs-canonical is this what you had in mind?

Batalex committed:
refactor: Separate refresh instantiation from usage in __init__

Batalex commented on src/charm.py:
```diff
@@ -225,6 +225,9 @@ def _collect_domain_statuses(self) -> list[ops.StatusBase]:  # noqa: C901 - igno
             statuses.append(Status.WAITING_FOR_SERVICE.value)
 
+        if not self.workload.active():
+            statuses.append(Status.WORKLOAD_NOT_RUNNING.value)
```
I just dropped this commit, I did not mean to push it on GH, I was cherry picking stuff locally and to figure out why the deployments were not starting the workload (fixed by moving the self.refresh assignment before the rest of the event handlers)

Batalex committed:
typo

Batalex committed:
fix: Resolve missing attr due to circular dep

Batalex committed:
fix: Restart workload post refresh

Batalex committed:
fix: Wrap refresh tests in tenacity retries

Batalex committed:
chore: Add ip log in the test

deusebio reviewed (approved):
All good! Thanks @Batalex ! Quite eager to see it in action!!!

welpaolo reviewed (approved):
Great work Alex!
